### ISSUE TITLE
Silm's PSR7 Response is incompatible with Guzzle's Stream

### ISSUE THREAD

**Issue by ckressibucher** [slimphp/Slim#1860]
When the response is cloned, the body (the guzzle stream in this case)
is cloned too. When the garbage collector clears the body instance
of the old response, the resource gets closed by the `__destruct`
method of guzzle. This causes unacceptable problems if the same
resource (the underlying resource of the guzzle body) is also
used in the new response (in its cloned body, to be precise).

Is there a reason to clone the body when cloning the response?

In my opinion, this is not a bug in guzzle. Closing the underlying
resource in the wrapping stream class's `__destruct` is fine.

**Comment by ckressibucher** [comment]
To show the problems, here is a script and its output:

```
<?php
require '../vendor/autoload.php';

class MyGuzzleStream extends GuzzleHttp\Psr7\Stream
{

    public function __destruct()
    {
        echo "\ndestructing guzzle stream " . spl_object_hash($this) . " ; following the stream resource which gets closed:\n";
        try {
            print_r($this->getMetadata());
        } catch (\Throwable $t) {
            echo PHP_EOL . $t->getMessage() . PHP_EOL;
        }
        parent::__destruct();
    }
}

class SlimResponse extends \Slim\Http\Response
{
    public function __destruct()
    {
        echo 'destructing ' . get_class($this) . ' hash ' . spl_object_hash($this) . "\n";
    }
}

$stream = new MyGuzzleStream(fopen('php://temp', 'r+'));
$response = new SlimResponse(200, null, $stream);

// copy (clone) response: Slim also clones the body stream
$respNew = $response->withHeader('Content-Type', 'text/plain');

echo "\nfinish... clear up resources...\n";
```

And the output:

```

finish... clear up resources...
destructing SlimResponse hash 0000000062e2bd95000000002bd2a9da

destructing guzzle stream 0000000062e2bd97000000002bd2a9da ; following the stream resource which gets closed:
Array
(
    [wrapper_type] => PHP
    [stream_type] => TEMP
    [mode] => w+b
    [unread_bytes] => 0
    [seekable] => 1
    [uri] => php://temp
)
destructing SlimResponse hash 0000000062e2bd93000000002bd2a9da

destructing guzzle stream 0000000062e2bd92000000002bd2a9da ; following the stream resource which gets closed:

Warning: stream_get_meta_data(): supplied resource is not a valid stream resource in /home/claudio/dev/php-qatrends/qatrends/vendor/guzzlehttp/psr7/src/Stream.php on line 236

Unsupported operand types
```

Note that we have 2 different guzzle streams, due to cloning. When the first instance is destroyed, it closes the stream (before that I return the metadata, which works). When the second instance is destroyed, getting the metadata does not work anymore, because the underlying stream has been closed before. This causes the `Error`.

**Comment by geggleto** [comment]
Actual answer to your question.

> Is there a reason to clone the body when cloning the response?
> PSR-7 dictates it must be cloned.

Ref: http://www.php-fig.org/psr/psr-7/#3-1-psr-http-message-messageinterface

> Messages are considered immutable; all methods that might change state MUST
> - be implemented such that they retain the internal state of the current
> - message and return an instance that contains the changed state.

**Now as to the why it behaves like it does...**

First off It's not the first Stream entry that get's deconstructed first... its actually the Last Stream.

Additional Debugging.
`debug_zval_dump();`

https://www.irccloud.com/pastebin/zNltCuWg/

I have added more detailed debug logs here.

The order of construction...
`
$stream = object(MyGuzzleStream)#2
$response = object(SlimResponse)#3 - > ref to  object(MyGuzzleStream)#2 
$respNew = object(SlimResponse)#5 ->  ref to object(MyGuzzleStream)#7
`

The order of the destructors is then...
`
object(SlimResponse)#5 
object(MyGuzzleStream)#7 
object(SlimResponse)#3
object(MyGuzzleStream)#2
`

So you can see it is in fact the last clone that gets cleaned up first. IIRC PHP does variable assignments like a stack, so it just unwinds from bottom to top the opposite to how the script executes. 

**All's okay.**
Moving the parent::__deconstruct to the top of the method...PHP-ISM
https://www.irccloud.com/pastebin/IK0CW1F3/

It looks fine to me.

**Comment by ckressibucher** [comment]
Thanks for testing.

PSR7 requires to clone the reponse when changing state of the response. As I understand it, it does not require a _deep_ clone. Actually, I didn't see it anywhere else...

Moving `parent::__destruct` to the top helps because guzzle then does not try to access the resource anymore -- it knows it's already outdated. My test script was written in a way that shows the problem arising from the fact that there are _2 different PSR7-Stream objects in memory, which use the same underlying php resource_. This is not an issue as long as nothing happens with the "old" object (the old response, in this case). But as soon as this old instance gets destructed, and in case of the guzzle Stream implementation, this causes to close the underlying stream, it also has an effect on the current `Response` object (which uses the same underlying stream). With moving the `parent::__destruct` call you "fixed" my test script, to not show the real world problem anymore, which still exists :-)

By the way, the problem I encounter in my real project is an endless loop when sending the response, because Slim checks for `eof` of the stream in `\Slim\App::respond`, and guzzle's implementation for this is: `return !$this->stream || feof($this->stream);`. If the `stream` (the underlying php resource) still exists, but has been closed by another, destructed, object, this returns `false` because `feof` fails (with a warning, because the stream has been closed). Slim is happy and continues to "read" from the stream (which kind of works, because guzzle does not check the state of the real resource, but its own `readable` property, which does not yet know the resource has been closed by another object). _Please do not waste time debugging this issue_, it's just a background info so you see how important it is to find a solution to the root cause for this problem. Which in my opinion is clearly, that it is possible to create 2 instances of `Stream` objects which use the same php resource.

**Comment by geggleto** [comment]
Does the problem exist if you use Slim's response object?

**Comment by geggleto** [comment]
I always thought it was kinda silly that we "cloned" the steam instead of opening a new stream and copying it.

**Comment by ckressibucher** [comment]
> Does the problem exist if you use Slim's response object?

Do you mean the Slim's Stream object?
When I use Slim Response and Slim Stream classes, I don't have this problems. Because the Slim Stream doesn't have a destructor which closes the stream resource.

> I always thought it was kinda silly that we "cloned" the steam instead of opening a new stream and copying it.

This would solve the issue, however, I'm not sure about the impact on performance...  If you have lots of data in the streams and you copy it over to a new one on every `withXYZ` method call... This is probably much worse than just instantiating new objects.

I don't understand the immutability required by the PSR spec as a deep immutability, so I think it would be ok to just copy over the reference of the Body Stream when cloning the Response object. All PSR7 implementations I saw until now do it like this.

**Comment by geggleto** [comment]
Well when you said...

> When I use Slim Response and Slim Stream classes, I don't have this problems. Because the Slim Stream doesn't have a destructor which closes the stream resource.

I don't really see how we can be of help. On our Response object we don't include a destructor. This issue should be raised to the Guzzle Project.

**Comment by akrabat** [comment]
I suspect that Slim is getting it wrong by cloning the StreamInterface object when the response is cloned. 

The underling PHP stream resource is not deep cloned when the StreamInterface object is clone and given that we can't assume that any given implementation has implemented stream_copy_to_stream() in their StreamInterface's __clone() method, Slim probably shouldn't be cloning the StreamInterface object.

i.e.it's my view that this is a Slim bug and that we should update `Slim\Http\Response::__clone()` to not clone the body.

Thoughts?

**Comment by akrabat** [comment]
@ckressibucher wrote:

> This would solve the issue, however, I'm not sure about the impact on performance... If you have lots of data in the streams and you copy it over to a new one on every withXYZ method call... This is probably much worse than just instantiating new objects

I agree. Copying the underlying data in the stream on every clone would be prohibitive.

Currently we're cloning the StreamInteface object, and so the underlying PHP stream is not copied. Hence the memory usage is negligible. You can see this with this code which creates a body of 1 million characters:

``` php
$app->get("/", function ($request, $response, $args) {
    $start = memory_get_usage();
    echo "<p>Initial memory: $start<br>";
    $response->write(str_pad("", 1000000, "x"));
    $second = memory_get_usage();
    echo "Memory after writing 1000000 characters: $second. (diff: " . ($second - $start) . ")<br>";
    $response = $response->withStatus(200);
    $third = memory_get_usage();
    echo "Memory after cloning: $third. (diff: " . ($third - $second) . ")<br>";

    return $response;
});
```

For me, the result is:

```
Initial memory: 1889000
Memory after writing 1000000 characters: 2893256. (diff: 1004256)
Memory after cloning: 2893632. (diff: 376)
```

**Comment by ckressibucher** [comment]
@geggleto The reason of psr standards is to ensure interoperability. So you _should_ care about compatibility with other implementations.

**Comment by ckressibucher** [comment]
@akrabat 

> and given that we can't assume that any given implementation has implemented stream_copy_to_stream() in their StreamInterface's __clone() method, Slim probably shouldn't be cloning the StreamInterface object.

Exactly. That's what I tried to explain. Regarding the `stream_copy_to_stream` inside the `StreamInterface` object's `__clone` method: I have not seen _any_ implementation so far that does it. (Slim does it in  Slim/Slim/Http/RequestBody.php but not in the response).

**Comment by akrabat** [comment]
Fixed in #1862
